Draft: Use std::variant for servo input pipeline

tylerjw opened the pull request:
### Description

Please explain the changes you made, including a reference to the related issue if applicable

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

felixvd committed:
Renaming, housekeeping, migration notes

Also added orientation checks

felixvd committed:
Unify object scaling in Rviz

felixvd committed:
Add pose to AttachedBody and RobotState

Co-authored by: JafarAbdi <cafer.abdi@gmail.com>

felixvd committed:
Add pose to CollisionObject and world

rhaschke committed:
Cleanup CollisionDetectorAllocatorTemplate::getName()

The change from NAME_ to getName() wasn't updated in the doc comment.
getName() doesn't need to be re-declared as abstract.

rhaschke committed:
Fixup devel-space build after #2604

gavanderhoorn committed:
controller manager: enclose name in quotes (#2761)

So empty controller names are easier to spot.

v4hn committed:
Merge pull request #2758 from ubi-agni/master

Cleanup CollisionDetectorAllocatorTemplate::getName()
AND
fixup cmake-generated declarations in devel workspaces

rhaschke committed:
CI: Switch to Noetic builds by default (#2765)

+ Enforce BUILDER=catkin_tools for Noetic

rhaschke committed:
Use (non-isolated) devel-space build

Isolated devel spaces fail due to dependency issues!

rhaschke committed:
Invalidate workspace caches on changes of ci.yaml

rhaschke committed:
Upload failures of different jobs separately

rhaschke committed:
Fix MSA templates (#2769)

* create static_transform_publisher for each virtual joint type

* another $(dirname)

augmenting #2748

* formatting

rhaschke committed:
Revert $(dirname) use for rviz config file

... due to this bug: https://github.com/ros/ros_comm/issues/1487
This partially reverts 442c3202a4124877afbb6e2bdee682c537f25553

rhaschke committed:
Fix Windows CI (#2776)

Fix error C2338: You've instantiated std::aligned_storage<Len, Align>
with an extended alignment (in other words, Align > alignof(max_align_t)).
Before VS 2017 15.8, the member type would non-conformingly have an
alignment of only alignof(max_align_t).
VS 2017 15.8 was fixed to handle this correctly, but the fix inherently
changes layout and breaks binary compatibility (*only* for uses of
aligned_storage with extended alignments). Please define either
(1) _ENABLE_EXTENDED_ALIGNED_STORAGE to acknowledge that you understand
    this message and that you actually want a type with an extended alignment
(2) _DISABLE_EXTENDED_ALIGNED_STORAGE to silence this message and
    get the old non-conformant behavior.

pvanlaar committed:
clang-tidy: modernize-make-shared, modernize-make-unique (#2762)

Co-authored-by: Robert Haschke <rhaschke@techfak.uni-bielefeld.de>

mathias-luedtke committed:
Add missing dependencies to generated dynamic_reconfigure headers

rhaschke committed:
MGI::setStartState: Only fetch current state when new state is diff (#2775)

v4hn committed:
Trajectory: Improve docstrings (#2781)

kevinshchang committed:
fix trajectory constraints for moveit commander (#2429)

remove unimplemented ability to load trajectory constraints from str

Fixup by v4hn <me@v4hn.de>

rhaschke committed:
Merge #2772: CI Improvements

captain-yoshi committed:
Fix bullet plugin library path name (#2783)

v4hn committed:
Fix obvious typo (#2787)

v4hn committed:
work around clang diagnostics warning

We could disable the warning, but the fix is simple enough
and the warning might be useful in the future.

v4hn committed:
ompl: provide override for missing isValid method

missing this produced clang warnings because the additional isValid methods hide the non-overriden method.
Interesting case of API incompatibility (for `-Werror`) by **adding** a new method to an interface...

felixvd committed:
Add planning_pipeline_id to MotionSequence service (#2755)

* Add planning_pipeline_id to MotionSequence action and service

* check for empty request

Co-authored-by: Joachim Schleicher <j.schleicher@pilz.de>

v4hn committed:
Merge pull request #2802 from v4hn/pr-master-fix-clang-issues

Isolated fixes from https://github.com/ros-planning/moveit/pull/2792

werner291 committed:
Bugfix in RDFLoader (#2806)

v4hn committed:
fix clang-tidy CI job (#2792)

v4hn committed:
more fixes for the clang-tidy job (#2813)

tylerjw committed:
Use std::variant for servo input pipeline

tylerjw commented on moveit_ros/moveit_servo/include/moveit_servo/input_pipeline.hpp:
```diff
@@ -0,0 +1,85 @@
+ *
+ * @return     The input pipeline.
+ */
+template <typename Visitor>
+auto makeInputPipeline(const rclcpp::Node::SharedPtr& node, const ServoParameters& parameters,
+                       const std::function<void()>& halt, Visitor&& next)
+{
+  if (parameters.low_latency_mode)
+  {
+    return InputSubscriber{ node, parameters.cartesian_command_in_topic, parameters.joint_command_in_topic,
+                            std::move(next) };
+  }
+  else
+  {
+    auto stale_command_halt =
+        StaleCommandHalt{ node, rclcpp::Duration::from_seconds(parameters.incoming_command_timeout), halt,
+                          std::move(next) };
+    auto resampler = InputResampler{ node, parameters.publish_period, std::move(stale_command_halt) };
+    auto timestamp_now = TimestampNow{ node, std::move(resampler) };
+    return InputSubscriber{ node, parameters.cartesian_command_in_topic, parameters.joint_command_in_topic,
+                            std::move(timestamp_now) };
+  }
+}
```
@griswaldbrooks here is my latest struggle.  What is the output type of this?  How might I store the output of this without resorting to dynamic memory?  Is this where I should use another variant?  I know what one of the output types is: `InputSubscriber<TimestampNow>`, but I do not know the output type of the other yet.  If possible I'd like to avoid putting that type in here because it is `ServoCalcs` which would make this much more difficult to test (it is not simple to mock/fake for testing).

For some context, much of the logic in this PR is currently buried in `ServoCalcs`.  I have not started reworking that to hook this up to it yet, so far I've just been building the various pieces and writing tests for them to verify the interfaces.

Another thing I'd like your critical review of is how I'm dealing with ownership of the Visitor objects.  As it is I'm giving ownership of the visitor object to the object that calls visit on it.  This simplifies dealing with the lifetime of each object.  I don't have to have some sort of container object for each step to manage their lifetime.  That is especially nice because I want to define which ones I construct at run time.

You'll also notice this is all implemented in headers so far.  That is just out of my own laziness as it is easier to write this way for me until I am sure of my interfaces and then I am happy to go move whatever implementation I can to source files to speed up compiling.

griswaldbrooks replied:
Do you need to specify a return type? `auto` works in C++17. I'm pretty sure the copy is elided on the return here so you don't have to use `unique_ptr`, though I might leave the side-effect object out of it and build it more like
```
auto sub = InputSubscriber{node,
                           parameters.cartesian_command_in_topic,
                           parameters.joint_command_in_topic,
                           makeInputPipeline(node, parameters, halter, []{})};
```
or some other factory for the `InputSubscriber` so you have more freedom to build and move pipeline objects.

griswaldbrooks replied:
Also, if it made sense, I obviously would try to find a way to DI that `node` parameter out of things

griswaldbrooks replied:
wrt visitor ownership, you don't have any side effects in their ctors, so I wouldn't do anything clever.

tylerjw committed:
Does not build, test for compiling InputPipeline builder

tylerjw committed:
Use non-owning raw pointers to visitor

schornakj committed:
Update controller_manager_plugin to fix MoveIt-managed controller switching (#785)

* Update controller_manager_plugin to fix MoveIt-managed controller switching

* add documentation about mutex

* add a comment about how interface names are modified in-place

Abishalini committed:
Merge https://github.com/ros-planning/moveit/commit/bec399ae8a64ac476d44c31400474aa55649db72

tylerjw committed:
Add backward_ros to moveit_common (#794)

Co-authored-by: Henning Kayser <henningkayser@picknik.ai>
Co-authored-by: Michael Görner <me@v4hn.de>

Abishalini committed:
Merge https://github.com/ros-planning/moveit/commit/779b7c8b019f70898d4de3189f9261c9697d9b9f

tylerjw committed:
Merge ros-planning/moveit@779b7c8

Abishalini committed:
Tests for TrajectoryMonitor using dependency injection (#570)

Co-authored-by: Tyler Weaver <tyler@picknik.ai>

davetcoleman committed:
Update Maintainers of MoveIt package (#697)

Co-authored-by: Tyler Weaver <tyler@picknik.ai>
Co-authored-by: Henning Kayser <henningkayser@picknik.ai>

henningkayser committed:
Enforce package.xml format 3 Schema (#779)

schornakj committed:
Fix installation of moveit_ros_control_interface header files (#789)

Abishalini committed:
Merge https://github.com/ros-planning/moveit/commit/11c7bd7dc37035e090f0b3f31c5896b0eba09509

Commit:
Remove backport CI job, in favor of Mergify (#804)

henningkayser committed:
Clang-tidy fixes (#596)

stephanie-eng committed:
Update README (#812)

schornakj committed:
Make controller management logic more tolerant of missing or late ros2_control nodes (#792)

* allow handle instantiation if action server does not exist

* reload controller info before ensuring active controllers

* make isConnected protected

change MoveItSimpleControllerManager to always create new controller handles even if their associated action servers are not available

* add more documentation for ActionBasedControllerHandle

* fix string concatenation to appease clang-tidy

* remove node_ member variable and add clarifying comment

* add a more useful parameter name concatenation function

stephanie-eng committed:
Add descriptions and default values to servo parameters (#799)

Co-authored-by: Tyler Weaver <tylerjw@gmail.com>

Abishalini committed:
Merge https://github.com/ros-planning/moveit/commit/b57628ff75e25b2aba2bbe408c1250287199568b

henningkayser committed:
List Noetic MIGRATION notes under Rolling

tylerjw committed:
Hook up to ServoCalcs

tylerjw committed:
Wait for complete state
